### PR TITLE
[STORM-2082][SQL] add sql external module storm-sql-hdfs

### PR DESCRIPTION
**Pull request by vesense**

### CODE REVIEW

File: external/sql/storm-sql-external/storm-sql-hdfs/src/test/org/apache/storm/sql/hdfs/TestHdfsDataSourcesProvider.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    FSDataOutputStream out = mock(FSDataOutputStream.class);
+    Whitebox.setInternalState(state, "options", options);
+    Whitebox.setInternalState(options, "out", out);
+
+    List<TridentTuple> tupleList = mockTupleList();
+
+    for (TridentTuple t : tupleList) {
+      stateUpdater.updateState(state, Collections.singletonList(t), null);
+      try {
+        verify(out).write(argThat(new HdfsArgMatcher(t)));
```
Comment (HeartSaVioR) [inline external/sql/storm-sql-external/storm-sql-hdfs/src/test/org/apache/storm/sql/hdfs/TestHdfsDataSourcesProvider.java]: @vesense 
testHdfsSink() is failing consistency. I guess options are mocked so options.execute() doesn't work as expected. You need to set options.execute() to run real method, and provide other needed fields as well.

File: external/sql/storm-sql-external/storm-sql-hdfs/src/jvm/org/apache/storm/sql/hdfs/HdfsDataSourcesProvider.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+          .withName(props.getProperty("hdfs.file.name", "$TIME.$NUM.txt"));
+
+      RecordFormat recordFormat = new TridentRecordFormat(serializer);
+
+      FileRotationPolicy rotationPolicy;
+      String size = props.getProperty("hdfs.rotation.size.kb");
+      if (size != null) {
+        rotationPolicy = new FileSizeRotationPolicy(Float.parseFloat(size), FileSizeRotationPolicy.Units.KB);
+      } else {
+        float interval = Float.parseFloat(props.getProperty("hdfs.rotation.time.seconds", "600")); // default 600 seconds
```
Comment (vesense) [inline external/sql/storm-sql-external/storm-sql-hdfs/src/jvm/org/apache/storm/sql/hdfs/HdfsDataSourcesProvider.java]: No default value used from storm-hdfs. I will add a preconditions check.

File: external/sql/storm-sql-external/storm-sql-hdfs/src/test/org/apache/storm/sql/hdfs/TestHdfsDataSourcesProvider.java (hunk 1 above)
Comment (vesense) [inline external/sql/storm-sql-external/storm-sql-hdfs/src/test/org/apache/storm/sql/hdfs/TestHdfsDataSourcesProvider.java]: Will fix.

**Comment by vesense** [comment]
@HeartSaVioR Addressed.

**Comment by HeartSaVioR** [comment]
@vesense 
Please provide `working SQL example statements` to test, and how to inject Hadoop resource files.
And please write up content for STORM-2202 (#1777).

**Comment by vesense** [comment]
@HeartSaVioR Now I'm a little busy, I will add it later today.

**Comment by HeartSaVioR** [comment]
No problem, please take your time.

**Comment by vesense** [comment]
>provide `working SQL example statements` to test

Kafka -> HDFS

```
CREATE EXTERNAL TABLE ORDERS (ID INT PRIMARY KEY, UNIT_PRICE INT, QUANTITY INT) LOCATION 'kafka://localhost:2181/brokers?topic=orders' TBLPROPERTIES '{"producer":{"bootstrap.servers":"localhost:9092","acks":"1","key.serializer":"org.apache.org.apache.storm.kafka.IntSerializer","value.serializer":"org.apache.org.apache.storm.kafka.ByteBufferSerializer"}}'
CREATE EXTERNAL TABLE LARGE_ORDERS (ID INT PRIMARY KEY, TOTAL INT) LOCATION 'hdfs://localhost:9000/' TBLPROPERTIES '{"hdfs.file.path":"/unittest","hdfs.file.name":"test-$NUM.txt","hdfs.rotation.time.seconds":"120"}'
INSERT INTO LARGE_ORDERS SELECT ID, UNIT_PRICE * QUANTITY AS TOTAL FROM ORDERS WHERE UNIT_PRICE * QUANTITY > 50
```

>how to inject Hadoop resource files

Put required jars into the `extlib` directory and configurations(core-site.xml and hdfs-site.xml) into the `conf` directory.

>write up content for STORM-2202

Will update.

File: docs/storm-sql-reference.md [hunk 3]
```diff
@@ -1312,4 +1313,23 @@ You can use below as working reference for `--artifacts` option, and change depe
...
+
+Please note that `hdfs.rotation.size.kb` and `hdfs.rotation.time.seconds` only one can be used for hdfs rotation.
+
+And note that `storm-sql-hdfs` requires users to provide `storm-hdfs`.
+You can use below as working reference for `--artifacts` option, and change dependencies version if really needed:
+
+`org.apache.storm:storm-sql-hdfs:2.0.0-SNAPSHOT,org.apache.storm:storm-hdfs:2.0.0-SNAPSHOT`
+
+Also, hdfs configuration files should be provided.
+You can put the `core-site.xml` and `hdfs-site.xm` into the `conf` directory.
```
Comment (HeartSaVioR) [inline docs/storm-sql-reference.md]: typo: `hdfs-site.xm` -> `hdfs-site.xml`.
and do you mean putting files to `conf` dir in Storm installation directory? Could you explain how they can be referred from workers?

Comment (vesense) [inline docs/storm-sql-reference.md]: Will fix typo.
Yes the `conf` dir is the one in Storm installation directory. The `conf` dir is added to classpath in supervisor `BasicContainer`. see: https://github.com/apache/storm/blob/2832aeb369d431a2e2f0709175db68f079855387/storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java#L350

Maybe there is a better way to locate the hdfs configuration files. For now users have to put them in each supervisor node.

**Comment by vesense** [comment]
@HeartSaVioR Updated and Squashed.

**Comment by HeartSaVioR** [comment]
Thanks for the update. Makes sense.
+1

**Comment by HeartSaVioR** [comment]
Also confirmed it works. I'll merge this.

**Pull request merged**